Issue 14736 tls ipv6 address

mattiash opened the pull request:
Resolves issue https://github.com/nodejs/node/issues/14736.
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
tls

mscdex commented on lib/_http_agent.js:
```diff
@@ -152,7 +152,12 @@ Agent.prototype.addRequest = function addRequest(req, options, port/*legacy*/,
+      // abc:123 => abc
+      // [::1] => ::1
+      // [::1]:123 => ::1
+      options.servername = hostHeader.replace(/:[^\]]+$/, '')
```
Why not combine both regexps for efficiency?:

```js
options.servername = hostHeader.replace(/^(?:\[([^\]]*)\].*)|(?:([^:]*).*)$/, '$1$2');
```

I tried a few different variations on the above regexps, and that one seemed to execute the fastest, at least on node v8.3.0.

mattiash replied:
I ran a benchmark on your regexp (regexp1), my regexp (regexp2) and a variant that doesn't use regexps (t3). See https://gist.github.com/mattiash/f19f7d862dce54db955267e03319b2d2

The result from my benchmark was that the two regexp-solution was actually faster than your one-regexp-solution, but the fastest was my new variant without any regexp.

```
$ node --version
v8.3.0
$ node index.js
regexp1 x 360,816 ops/sec ±1.15% (88 runs sampled)
regexp2 x 514,166 ops/sec ±1.21% (87 runs sampled)
t3 x 1,195,766 ops/sec ±1.38% (87 runs sampled)
Fastest is t3
```

```
$ node --version
v6.3.1
$ node index.js
regexp1 x 592,158 ops/sec ±0.95% (84 runs sampled)
regexp2 x 758,064 ops/sec ±1.21% (87 runs sampled)
t3 x 829,918 ops/sec ±1.18% (85 runs sampled)
Fastest is t3
``` 

I'll update the PR to use the t3 variant

jasnell commented on lib/tls.js:
```diff
@@ -163,6 +163,33 @@ function check(hostParts, pattern, wildcards) {
   return true;
 }
 
+exports._canonicalIp = function(address) {
```
I'd generally prefer not introducing new `_`-prefixed methods. Non-public methods should be moved into `lib/internal/` and used from there.

mattiash replied:
Fixed.

mattiash commented:
Is there anything else I need to do before this can be merged? We are using TLS with certificates that sign IPv4 addresses today, and we are very eager to add support for IPv6 which requires support for TLS with certificates that sign IPv6 addresses.

jasnell commented:
It needs a bit of additional review. I've added a couple of folks to the "Reviewers" list.

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/9802/

BridgeAR commented on lib/internal/tls.js:
```diff
@@ -0,0 +1,34 @@
+    const b = ['0', '0', '0', '0', '0', '0', '0', '0'];
+
+    const s = address.split('::');
+    if (s.length === 2) {
```
Is there actually any IPv6 without the two dots?

mattiash replied:
Yes. The :: notation is just an abbreviation for any number of zero-bytes. There can very well be an IPv6-address without any zero-bytes. And the address comes from the Host-header, so it might be that the http-client requested an address with the IPv6 address spelled out, e.g. https://[0.0.0.0.0.0.0.1] for localhost.

BridgeAR commented on lib/internal/tls.js:
```diff
@@ -0,0 +1,34 @@
+      }
+    }
+
+    return b.join(':');
```
If you want to improve the performance you could do the following to improve it further.
 - use string concatenation and not use the `b` array at all. And at the end add `"0:".repeat(restLen)`.
 - replace the RegExp with a loop that checks something like `var n = 0; while (str.charCodeAt(n++) === 48);` so you know what part you want to slice out.
 - use `address.indexOf('::')` and `address.split(':')` directly. You know until where the first part and until where the second part will go because of the indexOf and you only have to split a single time instead of three times (you could optimize this the same as the RegExp by using a loop and not splitting at all. It should be even faster).
- The `if (s1[n])` is not necessary as far as I see it. The current test cases have no entry that does not match but in case e.g. `:fe80::0000:` is a valid entry, just check the very first/last entry if it is entry and count down the length.

But this is out of my perspective not that important as a first draft and should not hold off the PR from being merged.

mattiash replied:
This code is run once per https-request, and only if the host is specified with an IPv6 address.  The code is already quite complex as it is and I would not like to complicate it further right now.

The `if (s1[n])` is needed for '::1'

BridgeAR commented on test/internet/test-tls-canonical-ip.js:
```diff
@@ -0,0 +1,18 @@
+assert.strictEqual(tls.canonicalIp('fe80::'), 'fe80:0:0:0:0:0:0:0');
+assert.strictEqual(
+  tls.canonicalIp('fe80::0000:0010:0001'),
+  'fe80:0:0:0:0:0:10:1');
```
It would be good to have some more test cases if there are any more out of the ordinary ones. It has been a long time that I checked how a valid IPv6 looks like.

mattiash replied:
Your comment made me realize that I didn't test IPv6-addresses without double colons. And of course that didn't work. Fix coming up.

BridgeAR commented:
PTAL @nodejs/http @nodejs/crypto

BridgeAR commented:
This needs some reviews @nodejs/collaborators

joyeecheung commented on test/internet/test-tls-canonical-ip.js:
```diff
@@ -0,0 +1,29 @@
+// of a TLS certificate
+
+const assert = require('assert');
+const tls = require('internal/tls');
```
Does this test work without `// Flags: --expose-internals`? (See [the testing guide](https://github.com/nodejs/node/blob/master/doc/guides/writing-tests.md#flags))

mattiash replied:
You're right, it actually doesn't. I have added that flag now. While debugging this I looked at the other tests in the test/internet/ directory and realised that they all needed internet access (which my test doesn't). I therefore moved the test to test/parallel which should make sure that it gets run by `make test`

BridgeAR commented:
@nodejs/tsc PTAL. 

@mattiash this needs a rebase.

mattiash commented:
@BridgeAR Should I rebase it now before the review is done, or will that interrupt the review process?

BridgeAR commented:
@mattiash it is always best to rebase and to have a clean state to review. Otherwise changed could come in be missed afterwards.

mattiash commented:
Branch rebased as requested. The only conflict was that someone had created internal/tls.js (which I also had) so I did a trivial merge. make -j4 test still passes for me locally.

mscdex commented:
FWIW it looks like there's something up with the author part of the commits' metadata. Perhaps there is a difference in author and committer email addresses or something?

mattiash commented:
> FWIW it looks like there's something up with the author part of the commits' metadata. Perhaps there is a difference in author and committer email addresses or something?

Yes, I wrote the patch at work and rebased it from home, that's why the author/committer don't match. I'll rebase again to avoid any confusion. Sorry for the trouble.

BridgeAR commented:
@nodejs/http @nodejs/crypto PTAL. This needs some LGs

mattiash commented:
Note to reviewers: https://github.com/mattiash/test-checkServerIdentity explains what the problem is and contains a test-case that fails with node 8.5.0 but succeeds with the code from this PR.

jasnell commented:
I have to agree with @bnoordhuis. Reusing one of the existing ipv6 parsing routines is best.

mattiash commented:
Thanks for the review. I looked at the different IPv6 parsing routines already in node, but I couldn't find anything that I could use out of the box to convert IPv6 addresses to a canonical format. I used @bnoordhuis suggestion about uv_inet_ntop/pton which works for my purposes.

trevnorris commented on src/cares_wrap.cc:
```diff
@@ -1935,6 +1935,32 @@ void IsIPv6(const FunctionCallbackInfo<Value>& args) {
+
+    args.GetReturnValue().Set(String::NewFromUtf8(isolate, canonical_ip));
+  } else {
+    args.GetReturnValue().Set(Undefined(isolate));
```
Default return value is `undefined`. No reason to set it here if it hasn't be set prior.

mattiash commented:
> Default return value is undefined. No reason to set it here if it hasn't be set prior.

Fixed.

bnoordhuis commented on lib/_http_agent.js:
```diff
@@ -152,7 +152,15 @@ Agent.prototype.addRequest = function addRequest(req, options, port/*legacy*/,
+      // [::1] => ::1
+      // [::1]:123 => ::1
+      if (hostHeader.startsWith('[')) {
+        options.servername = hostHeader.split(']')[0].substr(1);
```
`.split(']', 1)` is more efficient.

For bonus performance points, if you're going to slice the string anyway, call `.indexOf()` and then `.substr(1, index)` (and default `index` to `hostHeader.length` if it's -1.)

Lastly, I would say that it should pass on the string untouched if it doesn't contain a matching `]`.

bnoordhuis commented on src/cares_wrap.cc:
```diff
@@ -1935,6 +1935,30 @@ void IsIPv6(const FunctionCallbackInfo<Value>& args) {
   }
 }
 
+void canonicalIP(const FunctionCallbackInfo<Value>& args) {
```
Style: `CanonicalIP`

As well, I'd call it `CanonicalizeIP` for two reasons:

1. It makes it clear it's an imperative, not a predicate (i.e., it transforms, it doesn't just check.)
2. `lI` is almost indistinguishable from `ll`, `II`, etc. with some fonts.  (Minor thing.  Still.)

bnoordhuis commented on src/cares_wrap.cc:
```diff
@@ -1935,6 +1935,30 @@ void IsIPv6(const FunctionCallbackInfo<Value>& args) {
+    char canonical_ip[INET_ADDRSTRLEN];
+
+    int err = uv_inet_ntop(AF_INET, address_buffer, canonical_ip,
+      sizeof(canonical_ip));
```
Style: please line up the argument.

bnoordhuis commented on src/cares_wrap.cc:
```diff
@@ -1935,6 +1935,30 @@ void IsIPv6(const FunctionCallbackInfo<Value>& args) {
+void canonicalIP(const FunctionCallbackInfo<Value>& args) {
+  node::Utf8Value ip(args.GetIsolate(), args[0]);
+  char address_buffer[sizeof(struct in6_addr)];
+  v8::Isolate* isolate = args.GetIsolate();
```
Can you move this up and use it for `ip`?

bnoordhuis commented on src/cares_wrap.cc:
```diff
@@ -1935,6 +1935,30 @@ void IsIPv6(const FunctionCallbackInfo<Value>& args) {
+
+    args.GetReturnValue().Set(String::NewFromUtf8(isolate, canonical_ip));
+  } else if (uv_inet_pton(AF_INET6, *ip, &address_buffer) == 0) {
+    char canonical_ip[INET6_ADDRSTRLEN];
```
I'd just move this up a scope, it's big enough to hold IPv4 and IPv6 address strings.  Shaves off a few lines of code.

bnoordhuis commented on test/parallel/test-tls-canonical-ip.js:
```diff
@@ -0,0 +1,38 @@
+assert.strictEqual(canonicalIP('fe80:0:0:0:0:0:0:0'), 'fe80::');
+assert.strictEqual(
+  canonicalIP('fe80::0000:0010:0001'),
+  'fe80::10:1');
```
This fits on one line.  If it didn't, house style is to write it like this:
```js
assert.strictEqual(canonicalIP('fe80::0000:0010:0001'),
                   'fe80::10:1');
```
Only if that wouldn't fit in 80 columns would you use one-argument-one-line style but then please indent by four spaces.  (Truth be told, I would have expected the linter to complain about that.)

mattiash commented:
Addressed all comments from @bnoordhuis.

bnoordhuis commented on lib/_http_agent.js:
```diff
@@ -152,7 +152,21 @@ Agent.prototype.addRequest = function addRequest(req, options, port/*legacy*/,
+          options.servername = hostHeader.substr(1, index - 1);
+        }
+      } else {
+        options.servername = hostHeader.split(':')[0];
```
Can you `.split(':', 1)` here?

bnoordhuis commented on src/cares_wrap.cc:
```diff
@@ -1935,6 +1935,27 @@ void IsIPv6(const FunctionCallbackInfo<Value>& args) {
 }
 
+void CanonicalizeIP(const FunctionCallbackInfo<Value>& args) {
+  v8::Isolate *isolate = args.GetIsolate();
```
Tiny style nit: star leans left, i.e., `v8::Isolate* isolate`

mattiash commented:
Anything else I can do to get this merged? The merge conflict is trivial (a require() has been removed on a line adjacent to a line that I changed), but I don't dare to change it in fear of disturbing the review process, unless someone tells me to fix it.

gibfahn commented:
>but I don't dare to change it in fear of disturbing the review process, unless someone tells me to fix it.

Rebasing shouldn't affect the review process, please do rebase and fix the conflict.

It looks like there aren't any outstanding issues, so given that it's been approved, and no-one is -1 on it, it should be good to land. Hopefully it'll get a couple more approvals in the next day or two and then land.

mattiash commented:
> @cjihrig There's no PR-URL on this commit message. (core-validate-commit FTW as usual!)

I thought the PR-URL was something that should be added when the commits are squashed and merged? It seems to say so in the collaborator guide:

https://github.com/nodejs/node/blob/master/COLLABORATOR_GUIDE.md#landing-pull-requests

All these commits need to be squashed into a single commit when merging IMO, but I don't want to squash the commits while people are still reviewing and commenting.

bnoordhuis commented:
@mattiash That comment somehow shows up here but it's on a commit from a different PR.  Probably a GitHub bug.

mattiash commented:
> @mattiash That comment somehow shows up here but it's on a commit from a different PR. Probably a GitHub bug.

I finally figured out why the comment showed up in this report. It is because @Trott force-pushed to master after I rebased my PR on top of what was then current master:

https://github.com/nodejs/node/pull/15745#issuecomment-334563292

It means that the commit that I rebased on top of, now looks to be a part of this PR, even though a slightly modified variant of that commit is already on master. That is why this PR now includes the commit "deps: upgrade libuv to 1.15.0".

<img width="782" alt="screen shot 2017-10-10 at 14 49 47" src="https://user-images.githubusercontent.com/588898/31387345-69296cf4-adca-11e7-9252-a8de37912286.png">

mattiash commented:
Can we please merge this now? We are using an earlier variant of this patch to setup clusters that communicate over tls over IPv6 between different Amazon regions where we use our own CA to be able to sign IPv6 addresses. It is a bit annoying to have to build our own nodejs just to be able to communicate with the cluster

joyeecheung commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/10586/

mattiash commented:
Unfortunately, while browsing through the node code I found another place where the servername is set without handling IPv6 properly. This code is only run under very specific circumstances it seems, since I have been running with my patches for a long time without running into the problem. The following patch fixes the problem. I have chosen to put the servername code in a separate function now to avoid code duplication. Should I push the fix to this PR or should I open a new PR instead?

```udiff
From 156c7334c71565ecdda64ef1e48eb5440f248dff Mon Sep 17 00:00:00 2001
From: Mattias Holmlund <mattias.holmlund@netinsight.net>
Date: Fri, 13 Oct 2017 17:07:44 +0200
Subject: [PATCH] http: set servername correctly in createSocket

---
 lib/_http_agent.js | 54 ++++++++++++++++++++++++++----------------------------
 1 file changed, 26 insertions(+), 28 deletions(-)

diff --git a/lib/_http_agent.js b/lib/_http_agent.js
index f9fe60c8a7..5a6169614a 100644
--- a/lib/_http_agent.js
+++ b/lib/_http_agent.js
@@ -153,27 +153,7 @@ Agent.prototype.addRequest = function addRequest(req, options, port/*legacy*/,
   if (options.socketPath)
     options.path = options.socketPath;
 
-  if (!options.servername) {
-    options.servername = options.host;
-    const hostHeader = req.getHeader('host');
-    if (hostHeader) {
-      // abc => abc
-      // abc:123 => abc
-      // [::1] => ::1
-      // [::1]:123 => ::1
-      if (hostHeader.startsWith('[')) {
-        const index = hostHeader.indexOf(']');
-        if (index === -1) {
-          // Leading '[', but no ']'. Need to do something...
-          options.serverName = hostHeader;
-        } else {
-          options.servername = hostHeader.substr(1, index - 1);
-        }
-      } else {
-        options.servername = hostHeader.split(':', 1)[0];
-      }
-    }
-  }
+  ensureServerName(options, req);
 
   var name = this.getName(options);
   if (!this.sockets[name]) {
@@ -221,13 +201,7 @@ Agent.prototype.createSocket = function createSocket(req, options, cb) {
   if (options.socketPath)
     options.path = options.socketPath;
 
-  if (!options.servername) {
-    options.servername = options.host;
-    const hostHeader = req.getHeader('host');
-    if (hostHeader) {
-      options.servername = hostHeader.replace(/:.*$/, '');
-    }
-  }
+  ensureServerName(options, req);
 
   var name = self.getName(options);
   options._agentKey = name;
@@ -255,6 +229,30 @@ Agent.prototype.createSocket = function createSocket(req, options, cb) {
   }
 };
 
+function ensureServerName(options, req) {
+  if (!options.servername) {
+    options.servername = options.host;
+    const hostHeader = req.getHeader('host');
+    if (hostHeader) {
+      // abc => abc
+      // abc:123 => abc
+      // [::1] => ::1
+      // [::1]:123 => ::1
+      if (hostHeader.startsWith('[')) {
+        const index = hostHeader.indexOf(']');
+        if (index === -1) {
+          // Leading '[', but no ']'. Need to do something...
+          options.serverName = hostHeader;
+        } else {
+          options.servername = hostHeader.substr(1, index - 1);
+        }
+      } else {
+        options.servername = hostHeader.split(':', 1)[0];
+      }
+    }
+  }
+}
+
 function installListeners(agent, s, options) {
   function onFree() {
     debug('CLIENT socket onFree');
-- 
2.11.0

```

bnoordhuis commented:
@mattiash I'd make that `function ensureServerName(hostHeader)` pure function (i.e., accept and return a string, don't mutate objects) but otherwise good idea, yes.

mattiasholmlund commented:
Is there anything else I can do to get this merged? It's been two and a half months since I opened this PR and the PR solves a bug where node does the wrong thing every time which makes it impossible to use IPv6 and TLS-certificates that sign IP addresses. It might not be the most common use-case, but it is important for us and all the other programming languages that I have tried (java and go) handles it correctly.

apapirovski commented on src/cares_wrap.cc:
```diff
@@ -1935,6 +1935,27 @@ void IsIPv6(const FunctionCallbackInfo<Value>& args) {
+                           sizeof(canonical_ip));
+    CHECK_EQ(err, 0);
+
+    args.GetReturnValue().Set(String::NewFromUtf8(isolate, canonical_ip));
```
This is a total nit:

The two if clauses seem almost identical other than AF_INET(6), could this be more like:

```cpp
  int af;
  if (uv_inet_pton(AF_INET, *ip, &address_buffer) == 0)
    af = AF_INET;
  else if (uv_inet_pton(AF_INET6, *ip, &address_buffer) == 0)
    af = AF_INET6;
  else
    return;

  int err = uv_inet_ntop(af, address_buffer, canonical_ip,
                         sizeof(canonical_ip));
  CHECK_EQ(err, 0);

  args.GetReturnValue().Set(String::NewFromUtf8(isolate, canonical_ip));
```

apapirovski commented on node.gyp:
```diff
@@ -119,6 +119,7 @@
       'lib/internal/readline.js',
       'lib/internal/repl.js',
       'lib/internal/socket_list.js',
+      'lib/internal/tls.js',
```
Is there a reason this needs to be here twice? (It seems to already exist on line 124.)

mattiasholmlund replied:
No, I think that is some sort of merge conflict.

apapirovski commented on test/parallel/test-tls-canonical-ip.js:
```diff
@@ -0,0 +1,31 @@
+// of a TLS certificate
+
+const assert = require('assert');
+const canonicalizeIP = process.binding('cares_wrap').canonicalizeIP;
```
Nit: it would be nice to use destructuring here since there was a PR recently to convert most of the code base to more consistently do that.

mattiasholmlund commented:
Thanks for the review. I applied all your suggestions.

apapirovski commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/11028/ (since the other link is dead now)

If the CI is green and there are no objections or changes requested over the weekend, I think we should be able to land this next week.

apapirovski commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/11146/ (the last one was quite red)

apapirovski commented:
Ok, failures unrelated. Landing this now.

apapirovski commented:
Landed in b6df87e1d482f18202586273200fd7a8e9bc2fcd

Congrats on becoming a Contributor @mattiasholmlund!

MylesBorins commented:
Letting this bake in Current a bit longer before landing into LTS

mattiasholmlund commented:
Has this baked in master for long enough now to be considered for the v8 lts? The PR was included in Version 9.1.0 2017-11-07, so it has been out in the wild for over two months now.

gibfahn commented:
>Has this baked in master for long enough now to be considered for the v8 lts?

I think so, I've landed it on v8.x-staging.

MylesBorins commented:
backport to v6.x-staging in 8e7ac25aa6
